packages/whenable/src/watch.js

```diff
@@ -0,0 +1,237 @@
+// @ts-check
+import { E } from '@endo/far';
+import { M } from '@endo/patterns';
+
+import { getWhenablePayload, unwrapPromise } from './whenable-utils.js';
+
+const { Fail } = assert;
+
+const { apply } = Reflect;
+
+export const PromiseWatcherI = M.interface('PromiseWatcher', {
+  onFulfilled: M.call(M.any()).rest(M.any()).returns(),
+  onRejected: M.call(M.any()).rest(M.any()).returns(),
+});
+
+/**
+ * @typedef {object} PromiseWatcher
+ * @property {(...args: unknown[]) => void} [onFulfilled]
+ * @property {(...args: unknown[]) => void} [onRejected]
+ */
+
+/**
+ * Adapt a promise watcher method to E.when.
+ * @param {Record<PropertyKey, (...args: unknown[]) => unknown>} that
+ * @param {PropertyKey} prop
+ * @param {unknown[]} postArgs
+ */
+const callMeMaybe = (that, prop, postArgs) => {
+  const fn = that[prop];
+  if (!fn) {
+    return undefined;
+  }
+  assert.typeof(fn, 'function');
+  /**
+   * @param {unknown} arg value or reason
+   */
+  const wrapped = arg => {
+    // Don't return a value, to prevent E.when from subscribing to a resulting
+    // promise.
+    apply(fn, that, [arg, ...postArgs]);
+  };
+  return wrapped;
+};
+
+/**
+ * Shim the promise watcher behaviour when VatData.watchPromise is not available.
+ *
+ * @param {Promise<any>} p
+ * @param {PromiseWatcher} watcher
+ * @param {...unknown[]} watcherArgs
+ * @returns {void}
+ */
+const watchPromiseShim = (p, watcher, ...watcherArgs) => {
+  Promise.resolve(p) === p || Fail`watchPromise only watches promises`;
+  const onFulfilled = callMeMaybe(watcher, 'onFulfilled', watcherArgs);
+  const onRejected = callMeMaybe(watcher, 'onRejected', watcherArgs);
+  onFulfilled ||
+    onRejected ||
+    Fail`promise watcher must implement at least one handler method`;
+  void E.when(p, onFulfilled, onRejected);
+};
+
+/**
+ * @param {typeof watchPromiseShim} watchPromise
+ */
+const makeWatchWhenable =
+  watchPromise =>
+  /**
+   * @param {any} specimen
+   * @param {PromiseWatcher} promiseWatcher
+   */
+  (specimen, promiseWatcher) => {
+    let promise;
+    const payload = getWhenablePayload(specimen);
+    if (payload) {
+      promise = E(payload.whenableV0).shorten();
+    } else {
+      promise = E.resolve(specimen);
+    }
+    watchPromise(promise, promiseWatcher);
+  };
```

Comment:
Since I don't understand the purpose of the extra args, I am not suggesting the following. I'm just surprised it was not written this way and wondering: why?

```suggestion
/**
 * @param {typeof watchPromiseShim} watchPromise
 */
const makeWatchWhenable =
  watchPromise =>
  /**
   * @param {any} specimen
   * @param {PromiseWatcher} promiseWatcher
   * @param {...unknown[]} watcherArgs
   */
  (specimen, promiseWatcher, ...watcherArgs) => {
    let promise;
    const payload = getWhenablePayload(specimen);
    if (payload) {
      promise = E(payload.whenableV0).shorten();
    } else {
      promise = E.resolve(specimen);
    }
    watchPromise(promise, promiseWatcher, ...watcherArgs);
  };
```

Reply:
Resolving this along with the rest of my curiosity about these extra args. But I do remain curious!